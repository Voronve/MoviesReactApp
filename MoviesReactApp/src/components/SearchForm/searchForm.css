
  /* Style for the form */

  .searchContainer {
    text-align: left;
    padding: 100px 50px;
    background-color: black;
    background-blend-mode: multiply;
    background: rgba(0, 0, 0, 0.5);
    background-image: url('./images/Netflix.jpg');
  }

  .searchContainer h1 {
    font: 40px 'Montserrat', sans-serif;
    color: white;
    margin-bottom: 50px;
  }

  .searchForm {
    display: flex;
  }

  .searchFormElement {
    height: 57px;
    font-size: 20px;
    border-radius: 4px;
  }

  /* Style for the input field */
  .searchForm input {
    flex-grow: 1;
    padding-left: 8px;
    border: 1px solid #ccc;
    border-radius: 4px;
    outline: none;
    background-color: #323232;
    color: white;
    width: 713px;
    opacity: 70%;
    font-size: 20px;
  }

  .searchForm button {
    margin: 0;
    margin-left: 8px;
    background-color: #F65261;
    border: none;
    border-radius: 4px;
    color: white;
    cursor: pointer;
    min-width: 233px;
    height: inherit;
  }