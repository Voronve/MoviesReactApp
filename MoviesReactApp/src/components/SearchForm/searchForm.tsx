import React, { FormEvent, KeyboardEvent, ChangeEvent, useState } from 'react';
import './searchForm.css';
import { Link, Outlet, useSearchParams } from 'react-router-dom';
import { sortValue } from '../SortControl/sortControl';
import formQueryParamString from '../../helpers/formQueryParamsString';
export interface SearchFormProps {
/**
 * Initial search value
 */
 input: string,
/**
 * Callbeck function to be called on search
 */
 onSearch: (searchText: string) => void,
}

/** Form for searching movies */
export function SearchForm() {
    const [searchQuery, setSearchQuery] = useState('');
    const [searchParams, setSearchParams] = useSearchParams();
    const resultString = formQueryParamString(searchParams);

    const sortBy = searchParams.get('sortBy') as sortValue;
    const genre = searchParams.get('genre');

    const handleSubmit = (event: FormEvent<HTMLFormElement>) =>{
        event.preventDefault();
        const params = {} as {query: string, sortBy: sortValue, genre: string};
        if(searchQuery) params.query = searchQuery;
        if(sortBy) params.sortBy = sortBy;
        if(genre) params.genre = genre;

        setSearchParams(params);
    }

    const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {

        if (event.key === 'Enter') {
            event.preventDefault();
            const params = {} as {query: string, sortBy: sortValue, genre: string};
            if(searchQuery) params.query = searchQuery;
            if(sortBy) params.sortBy = sortBy;
            if(genre) params.genre = genre;
            setSearchParams(params);
        }
      };

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        setSearchQuery(event.currentTarget.value);
    };

    return (
        <div className="searchContainer">
            <h1>FIND YOUR MOVIE</h1>
            <form className="searchForm" onSubmit={handleSubmit}>
                <input
                    defaultValue={searchParams.get('query') || ''}
                    className="searchFormElement"
                    type="text"
                    placeholder="What do you want to watch?"
                    onChange={handleChange}
                    onKeyDown={handleKeyDown}/>
                <button className="searchFormElement" type="submit">Search</button>
            </form>
            <Link to={`/new${resultString}`}><div className='addMovie'>+ add movie</div></Link>
            <Outlet/>
        </div>
    );
}