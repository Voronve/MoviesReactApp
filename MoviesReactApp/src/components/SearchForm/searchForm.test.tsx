import { SearchForm } from './searchForm';
import { fireEvent, render, screen } from '@testing-library/react';

describe("Testing SearchForm component", () => {
    const input = 'Initial input';
    const onSearch = jest.fn();
    const customInput = "Horror";

    beforeEach(() => {
        render(<SearchForm/>);
    });

    test("Testing SearchForm input initial value", () => {

        fireEvent.click(screen.getByText('Search'));

        expect(onSearch).toBeCalledWith('Initial input');
    });

    test("Testing SearchForm input custom value after pressing \"Search\" button", () => {

        fireEvent.change(screen.getByRole("textbox"), {
            target: { value: customInput },
        });

        fireEvent.click(screen.getByText('Search'));

        expect(onSearch).toBeCalledWith(customInput);
    });

    test("Testing SearchForm input custom value after pressing Enter", () => {

        const textbox = screen.getByRole("textbox");

        fireEvent.change(textbox, {
            target: { value: customInput },
        });

        fireEvent.keyDown(textbox, {key: 'Enter', code: 'Enter', charCode: 13});

        expect(onSearch).toBeCalledWith(customInput);
    });
});