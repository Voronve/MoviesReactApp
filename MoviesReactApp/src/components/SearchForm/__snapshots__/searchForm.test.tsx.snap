// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchForm component Testing SearchForm rendering  1`] = `
<DocumentFragment>
  <div
    class="searchContainer"
  >
    <h1>
      FIND YOUR MOVIE
    </h1>
    <form
      class="searchForm"
    >
      <input
        class="searchFormElement"
        placeholder="What do you want to watch?"
        type="text"
        value=""
      />
      <button
        class="searchFormElement"
        type="submit"
      >
        Search
      </button>
    </form>
    <a
      href="/new"
    >
      <div
        class="addMovie"
      >
        + add movie
      </div>
    </a>
  </div>
</DocumentFragment>
`;
